apps/condo/domains/marketplace/schema/MarketCategory.test.js

```diff
@@ -0,0 +1,183 @@
+                expect(obj.updatedBy).toEqual(expect.objectContaining({ id: support.user.id }))
+            })
+
+            test('can\'t delete', async () => {
+                await expectToThrowAccessDeniedErrorToObj(async () => {
+                    await MarketCategory.delete(support, 'id')
+                })
+            })
+
+            test('can read', async () => {
+                const objs = await MarketCategory.getAll(support, {}, { sortBy: ['updatedAt_DESC'] })
+
+                expect(objs.length).toBeGreaterThanOrEqual(1)
+                expect(objs).toEqual(expect.arrayContaining([
+                    expect.objectContaining({
+                        id: checkReadAccessCategoryObj.id,
+                    }),
+                ]))
+            })
+        })
+
+        describe('user', () => {
+            test('can\'t create', async () => {
+                await expectToThrowAccessDeniedErrorToObj(async () => {
+                    await createTestMarketCategory(user)
+                })
+            })
+
+            test('can\'t update', async () => {
+                await expectToThrowAccessDeniedErrorToObj(async () => {
+                    await updateTestMarketCategory(user, 'id')
+                })
+            })
+
+            test('can\'t delete', async () => {
+                await expectToThrowAccessDeniedErrorToObj(async () => {
+                    await MarketCategory.delete(user, 'id')
+                })
+            })
+
+            test('can read', async () => {
+                const objs = await MarketCategory.getAll(user, {}, { sortBy: ['updatedAt_DESC'] })
+
+                expect(objs.length).toBeGreaterThanOrEqual(1)
+                expect(objs).toEqual(expect.arrayContaining([
+                    expect.objectContaining({
+                        id: checkReadAccessCategoryObj.id,
+                    }),
+                ]))
+            })
+
+        })
+
+        describe('anonymous', () => {
+            test('can\'t create', async () => {
+                await expectToThrowAuthenticationErrorToObj(async () => {
+                    await createTestMarketCategory(anonymous)
+                })
+            })
+
+            test('can\'t update', async () => {
+                await expectToThrowAuthenticationErrorToObj(async () => {
+                    await updateTestMarketCategory(anonymous, 'id')
+                })
+            })
+
+            test('can\'t delete', async () => {
+                await expectToThrowAccessDeniedErrorToObj(async () => {
+                    await MarketCategory.delete(anonymous, 'id')
+                })
+            })
+
+            test('can\'t read', async () => {
+                await expectToThrowAuthenticationErrorToObjects(async () => {
+                    await MarketCategory.getAll(anonymous, {}, { sortBy: ['updatedAt_DESC'] })
+                })
+            })
+        })
+
+    })
+
+    describe('logic tests', () => {
```

Comment:
Add a test for case when you indicate yourself as parent category (A -> A)

Reply:
I think it's internal_error

Reply:
such a case is only possible if nesting is greater than 2. So far we have a limit of 2.
A->B, B->A then when drawing the category tree there will be a recursion between these two categories